tensornetwork/network_operations.py

```diff
@@ -713,3 +713,28 @@ def contract_trace_edges(node: BaseNode) -> BaseNode:
     if edge.is_trace():
       return contract_parallel(edge)
   raise ValueError('`node` has no trace edges')
+
+
+def reduced_density(traced_out_edges: Iterable[Edge]) -> Tuple[dict, dict]:
+  """
+
+  Args:
+    traced_out_edges:
+
+  Returns:
+
+  """
+
+  # Get all reachable nodes.
+  old_nodes = get_all_nodes(traced_out_edges)
+
+  # Copy and conjugate all reachable nodes.
+  new_nodes, new_edges = copy(old_nodes, True)
+
+  # Add each edge to the copied nodes as new edge.
+  for traced_out_edge in traced_out_edges:
+    for new_node in new_nodes:
+      # What if a traced_out_edge is already an edge of the traget node?
+      # What axis should be set here axis1 or axis2?
+      # Do we need to update the nodes traced_out_edge connected to?
+      new_node.add_edge(traced_out_edge, traced_out_edge.axis1)
```

Comment:
Nope. Not quiet what you need to do.

I stead, just do 

```
for traced_out_edge in traced_out_edges:
  new_edge[traced_out_edge] ^ traced_out_edge
```

The `^` operation is the same as doing `tn.connect`

Reply:
Got it. Please take another look!

Reply:
One thing I am not sure is the dangling. It seems if the input is a connected graph. There will be chance that the ^ operation will fail when it tries to connect a traced_out_edge to its copy as the edge may not be dangling.

Reply:
Good observation!

Yes, `traced_out_edges` should only be dangling edges and we should raise a very clear error in the case that it's not. Can you add and test that as well?

Reply:
Done.